gui/mass-remove.lua

```diff
@@ -0,0 +1,350 @@
+-- building/construction mass removal/suspension tool
+
+--[====[
+
+gui/mass-remove
+===============
+Allows removal of buildings/constructions and suspend/unsuspend using
+a box selection.
+
+The following marking modes are available.
+
+- Suspend (s): suspends the construction of a planned building/construction
+- Unsuspend (p): resumes the construction of a planned building/construction
+- Remove Construction (n): designates a construction (wall, floor, etc) for removal. Similar to the native Designate->Remove Construction menu in DF
+- Unremove Construction (c): cancels removal of a construction (wall, floor, etc)
+- Remove Building (x): designates a building (door, workshop, etc) for removal. Similar to the native Set Building Tasks/Prefs->Remove Building menu in DF
+- Unremove Building (b): cancels removal of a building (door, workshop, etc)
+- Remove All (a): designates both constructions and buildings for removal, and deletes planned buildings/constructions
+- Unremove All (u): cancels removal designations for both constructions and buildings
+]====]
+
+local gui = require "gui"
+local guidm = require "gui.dwarfmode"
+local persistTable = require 'persist-table'
+local utils = require 'utils'
+local buildingplan = require('plugins.buildingplan')
+
+MassRemoveUI = defclass(MassRemoveUI, guidm.MenuOverlay)
+
+-- used to iterate through actions with + and -
+local actions={"suspend", "unsuspend", "remove_n", "unremove_n", "remove_x", "unremove_x", "remove_a", "unremove_a"}
+local action_indexes=utils.invert(actions)
+
+MassRemoveUI.ATTRS {
+    action="remove_a",
+    marking=false,
+    mark=nil
+}
+
+-- Helper functions.
+local function tablelength(T)
+  local count = 0
+  for _ in pairs(T) do count = count + 1 end
+  return count
+end
+
+-- Helper to match a job of a particular type at tile (x,y,z) and run the callback function on the job.
+local function iterateJobs(jobType, x, y, z, callback)
```

Comment:
One thing I missed earlier: I think `utils.listpairs()` would handle the loop for you. It's fairly simple logic, though, so I won't ask you to change it.